vncviewer/Viewport.cxx

```diff
@@ -1118,6 +1131,55 @@ int Viewport::handleSystemEvent(void *event, void *data)
         self->handleKeyRelease(0x36);
     }
 
+    return 1;
+  } else if ((msg->message == WM_CHAR) || (msg->message == WM_DEADCHAR) ||
+             (msg->message == WM_SYSCHAR) || (msg->message == WM_SYSDEADCHAR)) {
+    // Windows will send VK_PACKET key events if it doesn't have a physical key
+    // associated with the event (e.g. from a mobile device virtual keyboard).
+    // After translating a keydown/keyup event pair with VK_PACKET as vKey, a
+    // WM_*CHAR message will be issued containing a UTF-16 code unit. Unicode
+    // characters outside of the range U+0000 - U+ffff (Basic Multilingual
+    // Plane, BMP), are encoded as a pair of UTF-16 code units that will need to
+    // be synthesized into one Unicode code point.
+    uint32_t ucsCode = msg->wParam;
+    if ((ucsCode & 0xfc00) == 0xd800) {
```

Comment:
Could the surrogate handling be a separate commit? Makes it easier to see which parts are connected.

Reply:
I could, but only by making a first commit that exhibits wrong behavior w.r.t. surrogate pairs. Would you like that?